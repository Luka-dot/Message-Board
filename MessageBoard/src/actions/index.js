// actions
export const fetchNotes = (allNotes) => ({
  type: 'FETCH_NOTES',
  allNotes
})

export const hasError = (message) => ({
  type: 'HAS_ERROR',
  message
})

export const isLoading = (value) => ({
  type: 'IS_LOADING',
  value
}) 
