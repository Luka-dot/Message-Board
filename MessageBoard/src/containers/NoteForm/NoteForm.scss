.note-content {
  position: fixed;
  top: 50%; left: 50%;
  width: 450px; height: auto;
  transform: translate(-50%,-50%);
  box-shadow: 0px 2px 10px 3px #0000006b;
  display: flex;
  flex-direction: column;

  @media screen and (max-width: 500px) {
    width: 80%;
  }

  .note-title {
    font-family: $font-OpenSans;
    color: $textColor;
    font-size: 1.375rem;
    font-weight: 600;
    line-height: 1.75rem;
    min-height: 43px;
    padding: 7px 15px;
    margin-top: 3px;
    outline: none;
    border: none;
    background: inherit;
  }

  .errorLabel {
    display: none;
    align-items: center;
    padding: 0px 15px;
    font-size: .8em;

    &.Active {
      display: flex;
    }

    i {
      color: #DD1972;
      margin-right: .5em;
    }
    p {
      margin: 0;
      // font-size: .8em;
      color: #DD1972;
    }
  }

  .Completed-Message {
    min-height: 29px;
    padding: 5px 45px 5px 18px;
    text-align: left;

    p {
      margin: 0;
      font-size: .9em;
      color: $textColor;
      font-weight: 400;
      line-height: 1.5rem;
      letter-spacing: .00825em;
      font-family: $font-Roboto;
    }
  }

  .Note-Error {
    h2 { margin: 0; }
  }
}
