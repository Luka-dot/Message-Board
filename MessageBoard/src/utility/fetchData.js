export const fetchData = async (url, body) => {
  const response = await fetch(url, body)
  const data = await response.json();
  if(!response.ok){
    throw new Error(data)
  } else {
    return data;
  }
}