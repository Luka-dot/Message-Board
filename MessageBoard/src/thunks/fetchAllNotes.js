import { fetchNotes, hasError, isLoading } from '../actions';

export const fetchAllNotes = (url) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch(url);
      dispatch(isLoading(false));
      if(!response.ok){
        throw Error(response.statusText)
      }
      const result = await response.json();
      dispatch(fetchNotes(result));
    } catch (error) {
      dispatch(hasError(error.message));
    }
  }
}
