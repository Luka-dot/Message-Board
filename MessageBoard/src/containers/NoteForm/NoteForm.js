import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import shortid from 'shortid';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hasError, isLoading } from '../../actions/index';
import { fetchOptionsCreator } from '../../utility/fetchOptionsCreator'
import { fetchData } from '../../utility/fetchData';
import { fetchAllNotes } from '../../thunks/fetchAllNotes'
import { ListItem } from '../../components/ListItem/ListItem';
import NoteOptions from '../../components/NoteOptions/NoteOptions';

export class NoteForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      toHomePage: false,
      displayError: false,
      background: '#FFE680',
      list: [
        {
          id: shortid.generate(),
          isComplete: false,
          text: '',
        },
      ]
    }
  }

  componentDidMount = async () => {
    this.props.noteId && await this.findNote(this.props.noteId);
    document.addEventListener('keydown', this.handleKeydown);
    document.addEventListener('mousedown', this.handleClick);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKeydown);
    document.removeEventListener('mousedown', this.handleClick);
  }

  handleClick = (event) => {
    const noteClass = event.path.filter(road => road.className === 'Note');
    if (!noteClass.length) this.setState({toHomePage: true});
  }

  handleKeydown = (event) => {
    if ( event.code === 'Escape' ) this.setState({toHomePage: true});
  }

  findNote = async (noteId) => {
    this.props.isLoading(true);
    const url = `http://localhost:3001/api/v1/notes/${noteId}`;
    try {
      const response = await fetchData(url);
      const { id, title, list, background } = response;
      this.setState({
        id,
        background,
        title,
        list: [...list]
      })
    } catch(error) {
      if (error.message === 'Note was not found') {
        this.props.hasError(error.message);
        this.setState({displayError: true})
      }
    }
    this.props.isLoading(false);
  }

  handleSeparate = () => {
    const { list } = this.state;
    let completedItems = list.filter(item => item.isComplete);
    let incompletedItems = list.filter(item => !item.isComplete);
    return {
      completedItems,
      incompletedItems
    }
  }

  handleType = (e) => {
    e.preventDefault();
    const { type } = this.props;
    if (type === "new-note" || type === "existing-note") this.handlePostandPut(type)
  }

  handlePostandPut = async (type) => {
    const { title, list, background } = this.state;
    let path = type === 'new-note' ? 'POST' : 'PUT';
    let urlEnd = type === 'new-note' ? 'notes' : `notes/${this.props.id}`;
    const url = `http://localhost:3001/api/v1/${urlEnd}`;
    try {
      const options = await fetchOptionsCreator(path , { title, list, background });
      await fetchData(url, options);
      this.props.fetchAllNotes('http://localhost:3001/api/v1/notes');
      this.setState({toHomePage: true});
    } catch(error) {
      this.props.hasError('Note Cannot Be Saved');
      this.displayWindowAlert()
    }
  }

  handleTitleChange = (e) => {
    const { name, value } = e.target;
    if(value.length < 1) {
      this.setState({ displayError: true });
      this.setState({[name]: value});
    } else {
      this.setState({ displayError: false });
      this.setState({[name]: value});
    }
  }

  handleItemChange = (e, id) => {
    const { list } = this.state;
    const foundItem = list.find(item => item.id === id);
    foundItem.text = e.target.value;
    this.setState({ list });
    this.generateNewListItem(e, foundItem);
  }

  generateNewListItem = (e, foundItem) => {
    const { value } = e.target;
    const items = this.handleSeparate();
    const lastItem = items.incompletedItems.pop();
    if (value.length === 1 && foundItem.id === lastItem.id) this.addItem();
  }

  handleItemDelete = (e, id) => {
    e.preventDefault();
    const { list } = this.state;
    const foundIndex = list.findIndex(item => item.id === id);
    list.splice(foundIndex, 1);
    this.setState({ list });
  }

  addItem = () => {
    let defaultItem = {
      id: shortid.generate(),
      isComplete: false,
      text: ''
    };
    this.setState({list: [...this.state.list, defaultItem]});
  }

  deleteNote = async (e) => {
    e.preventDefault();
    const url = `http://localhost:3001/api/v1/notes`;
    try {
      const options = await fetchOptionsCreator('DELETE', {})
      await fetchData(`${url}/${this.props.id}`, options);
      this.props.fetchAllNotes(url);
      this.setState({ toHomePage: true });
    } catch (error) {
      this.props.hasError('Note Can Not Be Deleted');
      this.displayWindowAlert()
    }
  }

  displayWindowAlert = () => {
    window.alert(this.props.error + '\nInternal Server Error')
    this.setState({ toHomePage: true });
  }

  toggleComplete = (id) => {
    const { list } = this.state;
    const updatedList = list.map(item => {
      if (id === item.id) item.isComplete = !item.isComplete;
      return item;
    });
    this.setState({ list: updatedList });
  }

  changeColor = (e) => {
    const { name } = e.target;
    this.setState({ background: name })
  }

  handleClose = () => {
    this.props.hasError('');
  }

  render() {
    const separatedList = this.handleSeparate();
    const { completedItems, incompletedItems } = separatedList;
    const { toHomePage, displayError, background } = this.state;
    if (toHomePage) return <Redirect to='/' />;
    let completedMessage = completedItems.length ? `${completedItems.length} Completed Item(s)` : null;
    return (
      <div className="Note">
        {
          this.props.loading ?
            <div></div> :
            <div className="note-content" style={{backgroundColor: background}}>
              <input  type="text"
                className="note-title"
                placeholder="Enter A Title..."
                value={this.state.title}
                name="title"
                autoComplete="off"
                onChange={this.handleTitleChange} />
                <div className={displayError ? 'errorLabel Active' : 'errorLabel'}>
                  <i className="fas fa-exclamation-triangle"></i>
                  <p className="name-error">Please Enter A Title</p>
                </div>
              <ul className="ListItems">
                {
                  incompletedItems.map((item, index) =>
                    <ListItem key={item.id}
                              index={index}
                              toggleComplete={this.toggleComplete}
                              handleItemChange={this.handleItemChange}
                              handleItemDelete={this.handleItemDelete}
                              {...item} /> )
                }
              </ul>
              { completedMessage && <p className="Completed-Message">{completedMessage}</p> }
              <ul className="ListItems Completed">
                {
                  completedItems.map((item, index) =>
                    <ListItem key={item.id}
                              index={index}
                              toggleComplete={this.toggleComplete}
                              handleItemChange={this.handleItemChange}
                              handleItemDelete={this.handleItemDelete}
                              {...item} /> )
                }
              </ul>
              <NoteOptions  handleType={this.handleType}
                            deleteNote={this.deleteNote}
                            handleClose={this.handleClose}
                            type={this.props.type}
                            displayError={displayError}
                            changeColor={this.changeColor}
                            title={this.state.title}/>
            </div>
              }
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  error: state.error,
  loading: state.loading,
})

export const mapDispatchToProps = (dispatch) => ({
  hasError: (message) => dispatch(hasError(message)),
  fetchAllNotes: (url) => dispatch(fetchAllNotes(url)),
  isLoading: (value) => dispatch(isLoading(value)),
})

NoteForm.propTypes = {
  title: PropTypes.string,
  list: PropTypes.array,
  toHomePage: PropTypes.bool,
  error: PropTypes.string,
  fetchAllNotes: PropTypes.func,
  hasError: PropTypes.func
}

export default connect(mapStateToProps, mapDispatchToProps)(NoteForm)
