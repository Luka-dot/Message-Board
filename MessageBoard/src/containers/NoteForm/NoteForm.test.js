import React from 'react';
import { shallow } from 'enzyme';
import { NoteForm, mapStateToProps, mapDispatchToProps } from './NoteForm';
import { hasError } from '../../actions';

import { fetchData } from '../../utility/fetchData';
jest.mock('../../utility/fetchData')

import { fetchOptionsCreator } from '../../utility/fetchOptionsCreator';
jest.mock('../../utility/fetchOptionsCreator')

import { fetchAllNotes } from '../../thunks/fetchAllNotes';
jest.mock('../../thunks/fetchAllNotes');

import shortid from 'shortid';
jest.mock('shortid')
shortid.mockReturnValue("12345789")
window.alert = jest.fn()

let mockId = "O6oYExG7Q";
let noteId = "O6oYExG7Q";
let mockTitle = "Testing Note";
let mockType = "existing-note";


describe('NoteForm', () => {

  describe('NoteForm Component', () => {
    let wrapper;
    let mockfetchAllNotes = jest.fn()
    let mockHasError = jest.fn()
    let mockHasLoading = jest.fn()

    const mockNote = [
      {
        id: "987GDGFD",
        title: 'Jake ToDo',
        background: '#FFE680',
        list: [
        { id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false },
      ]
      }
    ];

    let mockFetchedNote = {
      id: "987GDGFD",
      title: 'Jake ToDo',
      background: '#FFE680',
      list: [
        { id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false },
      ]
    }

    beforeEach(() => {
      fetchData.mockImplementation(() => Promise.resolve(mockFetchedNote))

      wrapper = shallow(<NoteForm type={mockType}
                                  noteId={noteId}
                                  id={mockId}
                                  title={mockTitle}
                                  fetchAllNotes={mockfetchAllNotes}
                                  hasError={mockHasError}
                                  isLoading={mockHasLoading}
                        />)
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have the default state properties', () => {
      wrapper = shallow(<NoteForm people={['name']}/>, { disableLifecycleMethods: true })
      expect(wrapper.state()).toEqual({
        title: '',
        background: '#FFE680',
        list: [{
                id: "12345789",
                isComplete: false,
                text: ""
              }],
        toHomePage: false,
        displayError: false,
      })
    })

    it('should invoke findNote() if has a prop of noteId on componentDidMount', () => {
      const instance = wrapper.instance()
      jest.spyOn(instance, 'findNote')
      wrapper.instance().componentDidMount();
      expect(instance.findNote).toHaveBeenCalledWith(noteId)
    })

    it('should fire handleEscape when handleKeydown is invoked with the code is Escape', () => {
      const mockEvent = { code: 'Escape' };
      const instance = wrapper.instance();
      instance.handleKeydown(mockEvent);
      expect(wrapper.state('toHomePage')).toEqual(true);
    })

    it('should invoke fetchData and change the state to the mockData when findNote() is called', async () => {
      let mockUrl = "http://localhost:3001/api/v1/notes/O6oYExG7Q";

      fetchData.mockImplementationOnce(() => Promise.resolve(mockFetchedNote))
      await wrapper.instance().findNote();
      expect(fetchData).toHaveBeenCalledWith(mockUrl)
      expect(wrapper.state('title')).toEqual('Jake ToDo')
      expect(wrapper.state('list')).toEqual([{ id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false }])
    })

    it('should throw an error if fetchData is rejected in when the findNote() method is invoked and change its state', async () => {
      fetchData.mockImplementationOnce(() => Promise.reject(new Error('Note was not found')))

      await wrapper.instance().findNote()
      expect(mockHasError).toHaveBeenCalledWith('Note was not found')
      expect(wrapper.state('displayError')).toEqual(true)
    })

    it('should call handlePostandPut when handleType is invoked and has prop type of existing-note', () => {
      const instance = wrapper.instance();
      const fakeEvent = { preventDefault: () => {}};
      jest.spyOn(instance, 'handlePostandPut')
      wrapper.instance().handleType(fakeEvent)
      expect(instance.handlePostandPut).toHaveBeenCalledWith('existing-note')
    })

    it('should call handlePostandPut when handleType is invoked and has prop type of new-note', () => {
      const instance = wrapper.instance()
      const fakeEvent = { preventDefault: () => {} }
      jest.spyOn(instance, 'handlePostandPut');
      wrapper.setProps({ type: "new-note" })
      instance.handleType(fakeEvent)
      expect(instance.handlePostandPut).toHaveBeenCalledWith('new-note');
    })

    it('should call fetchOptionsCreator upon invoking handle post and return POST and title and list object', async () => {
      const title = 'Jake ToDo';
      const list = [{ id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false }]
      const background = '#FFE680';

      await wrapper.instance().handlePostandPut('new-note');
      expect(fetchOptionsCreator).toHaveBeenCalledWith('POST', { title, list, background })
    })

    it('shoud call fetchData when handlePostandPut is invoked with existing-note with the correct url and options', async () => {
      const mockUrl = 'http://localhost:3001/api/v1/notes/O6oYExG7Q';
      await wrapper.instance().handlePostandPut('existing-note')
      const mockOptions = await fetchOptionsCreator('PUT')
      expect(fetchData).toHaveBeenCalledWith(mockUrl, mockOptions)
    })

    it('shoud call fetchData when handlePostandPut is invoked with new-note with the correct url and options', async () => {
      const mockUrl = 'http://localhost:3001/api/v1/notes';
      await wrapper.instance().handlePostandPut('new-note')
      const mockOptions = await fetchOptionsCreator('POST')
      expect(fetchData).toHaveBeenCalledWith(mockUrl, mockOptions)
    })

    it('should call fetchAllNotes with the mockUrl when handlePostandPut is invoked and change the state of toHomePage', async () => {
      const mockUrl = 'http://localhost:3001/api/v1/notes';
      await wrapper.instance().handlePostandPut('new-note');

      expect(mockfetchAllNotes).toHaveBeenCalledWith(mockUrl)
      expect(wrapper.state('toHomePage')).toEqual(true)
    })

    it('should call hasError when handlePostandPut is invoked with a bad request', async () => {
      fetchData.mockImplementationOnce(() => Promise.reject(new Error('Note Cannot Be Saved')))
      const spy = spyOn(wrapper.instance(), 'displayWindowAlert')
      await wrapper.instance().handlePostandPut()
      expect(mockHasError).toHaveBeenCalledWith('Note Cannot Be Saved')
      expect(spy).toHaveBeenCalled()
    })

    it('should call fetchOptionsCreator with the correct params when handlePostandPut is invoked', async () => {
      const title = 'Jake ToDo';
      const list = [{ id: 'O6oYExG7Q', text: 'Do basic styling', isComplete: false }];
      const background = '#FFE680';

      await wrapper.instance().handlePostandPut()
      expect(fetchOptionsCreator).toHaveBeenCalledWith('PUT', { title, list, background })
    })

    it('should change state of title input when handleTitleChange is invoked', () => {
      const mockTitleEvent = { target: { name: 'title', value: 'This is Really Cool' } }
      wrapper.instance().handleTitleChange(mockTitleEvent)

      expect(wrapper.state('title')).toEqual('This is Really Cool')
    })

    it('should on invoking handleItemChange update the state of the list object to the new text and fire generateNewListItem with correct params', () => {
      wrapper.setState({
        list: [
          { id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false },
          { id: "55ogax554", text: 'Do More CsS', isComplete: true },
        ]
      })
      const mockFoundItem = { id: "55ogax554", isComplete: true, text: "This is updated list text"}
      const mockEvent = { target: { name: 'list', value: 'This is updated list text' }, preventDefault: () => {} }
      const instance = wrapper.instance()
      spyOn(instance, 'generateNewListItem')
      instance.handleItemChange(mockEvent, '55ogax554', 1)
      expect(wrapper.state('list')).toEqual([
        { id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false },
        { id: "55ogax554", text: 'This is updated list text', isComplete: true },
      ])
      expect(instance.generateNewListItem).toHaveBeenCalledWith(mockEvent, mockFoundItem)
    })

    it('should on invoking generateNewListItem invoke the method addItem', () => {
      wrapper.setState({
        list: [
          { id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false },
          { id: "55ogax554", text: 'Do More CsS', isComplete: true },
          { id: "1234552f4", text: 'Javascript is cool', isComplete: false },
        ]
      })
      const mockEvent = { target: { value: 'M'}, preventDefault: () => {}}
      const mockFoundItem = { id: "1234552f4", text: 'Javascript is cool', isComplete: false }
      const instance = wrapper.instance()
      spyOn(instance, 'addItem');
      instance.generateNewListItem(mockEvent, mockFoundItem)
      expect(instance.addItem).toHaveBeenCalled();
    })

    it('should on invoking addItem add a new list item to the original state of list', () => {
      const instance = wrapper.instance()
      instance.addItem()
      expect(wrapper.state('list')).toHaveLength(2)
      expect(wrapper.state('list')).toEqual(
        [{"id": "O6oYExG7Q", "isComplete": false, "text": "Do basic styling"}, {"id": "12345789", "isComplete": false, "text": ""}]
      )
    })

    it('should on invoking deleteNote call fetchOptionsCreator', async () => {
      const mockEvent = { preventDefault: () => {} }
      await wrapper.instance().deleteNote(mockEvent)
      expect(fetchOptionsCreator).toHaveBeenCalledWith('DELETE', {})
    })

    it('should on invoking deleteNote call fetchData with the correct parameters', async () => {
      const mockEvent = { preventDefault: () => {} }
      const mockUrl = 'http://localhost:3001/api/v1/notes/O6oYExG7Q';
      await wrapper.instance().deleteNote(mockEvent)
      expect(fetchData).toHaveBeenCalledWith(mockUrl)
    })

    it('should on invoking deleteNote call fetchallNotes with the correct url and set the state of toHomePage to true', async () => {
      const mockEvent = { preventDefault: () => {} }
      const mockUrl = 'http://localhost:3001/api/v1/notes';
      await wrapper.instance().deleteNote(mockEvent)
      expect(mockfetchAllNotes).toHaveBeenCalledWith(mockUrl)
    })

    it('should on invoking deleteNote with a bad request send the error correct message', async () => {
      fetchData.mockImplementationOnce(() => Promise.reject(new Error('Note not found')))
      const mockEvent = { preventDefault: () => {} }
      await wrapper.instance().deleteNote(mockEvent)
      expect(mockHasError).toHaveBeenCalledWith('Note was not found')
    })

    it('should on handleItemDelete splice the specific object out of the list array and then set state with the new info', () => {
      wrapper.setState({
        list: [
          { id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false },
          { id: "55ogax554", text: 'Do More CsS', isComplete: true },
          { id: "1234552f4", text: 'Javascript is cool', isComplete: false },
        ]
      })
      const mockListAfter = [{ id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false },
      { id: "55ogax554", text: 'Do More CsS', isComplete: true }]

      const mockEvent = { preventDefault: () => {} }
      const instance = wrapper.instance();
      instance.handleItemDelete(mockEvent, '1234552f4')
      expect(wrapper.state('list')).toEqual(mockListAfter)
    })

    it('should on toggleComplete switch the status of isComplete', () => {
      const updatedItem = [{ id: 'O6oYExG7Q', text: 'Do basic styling', isComplete: true }]
      const instance = wrapper.instance();
      instance.toggleComplete('O6oYExG7Q')
      expect(wrapper.state('list')).toEqual(updatedItem)
    })

    it('should on handleSeparate return an object with complete and uncompleted items', () => {
      wrapper.setState({
        list: [
          { id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false },
          { id: "55ogax554", text: 'Do More CsS', isComplete: true },
          { id: "1234552f4", text: 'Javascript is cool', isComplete: false },
        ]
      })
      const returnSeparate = {
        completedItems: [ { id: '55ogax554', text: 'Do More CsS', isComplete: true } ],
        incompletedItems: [
            { id: 'O6oYExG7Q', text: 'Do basic styling', isComplete: false },
            { id: '1234552f4', text: 'Javascript is cool', isComplete: false } ]
      }
      const instance = wrapper.instance()
      expect(instance.handleSeparate()).toEqual(returnSeparate)
    })

    it('should on handleClose call hasError with an empty string', () => {
      const instance = wrapper.instance()
      instance.handleClose()
      expect(mockHasError).toHaveBeenCalledWith('')
    })

    it('should update match the snapshot when there is more list items in the list state', () => {
      wrapper.setState({
        list: [
          { id: "O6oYExG7Q", text: 'Do basic styling', isComplete: false },
          { id: "55ogax554", text: 'Do More CsS', isComplete: true },
          { id: "1234552f4", text: 'Javascript is cool', isComplete: false },
        ]
      })
      expect(wrapper).toMatchSnapshot()
    })

    it('should update the background state with the name of the color', () => {
      const mockEvent = { target: { name: '#agedr44' } }
      wrapper.instance().changeColor(mockEvent)
      expect(wrapper.state('background')).toEqual('#agedr44')
    })

  })

  describe('mapStateToProps', () => {

    it('should return an object with the hasError message', () => {
      const mockState = {
        error: 'This is an Error Message',
        fitler: []
      }
      const expected = {
        error: 'This is an Error Message',
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected)
    })

  })


  describe('mapDispatchToProps', () => {

    it('should call dispatch with an hasError action when hasError is called', () => {
      // Setup
      const mockMessage = 'This is an error message';
      const mockDispatch = jest.fn();
      const actionToDispatch = hasError(mockMessage)

      // Execution
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.hasError(mockMessage)

      // Expectation
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should call dispatch with an fetchAllNotes action when fetchAllNotes is called', () => {
      // Setup
      const mockUrl = 'http://google.com'
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchAllNotes(mockUrl)

      // Execution
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.fetchAllNotes(actionToDispatch)

      // Expectaion
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

  })

})
