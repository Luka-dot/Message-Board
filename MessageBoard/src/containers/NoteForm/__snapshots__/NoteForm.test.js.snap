// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteForm NoteForm Component should match the snapshot 1`] = `
<div
  className="Note"
>
  <div
    className="note-content"
    style={
      Object {
        "backgroundColor": "#FFE680",
      }
    }
  >
    <input
      autoComplete="off"
      className="note-title"
      name="title"
      onChange={[Function]}
      placeholder="Enter A Title..."
      type="text"
      value="Jake ToDo"
    />
    <div
      className="errorLabel"
    >
      <i
        className="fas fa-exclamation-triangle"
      />
      <p
        className="name-error"
      >
        Please Enter A Title
      </p>
    </div>
    <ul
      className="ListItems"
    >
      <ListItem
        handleItemChange={[Function]}
        handleItemDelete={[Function]}
        id="O6oYExG7Q"
        index={0}
        isComplete={false}
        key="O6oYExG7Q"
        text="Do basic styling"
        toggleComplete={[Function]}
      />
    </ul>
    <ul
      className="ListItems Completed"
    />
    <NoteOptions
      changeColor={[Function]}
      deleteNote={[Function]}
      displayError={false}
      handleClose={[Function]}
      handleType={[Function]}
      title="Jake ToDo"
      type="existing-note"
    />
  </div>
</div>
`;

exports[`NoteForm NoteForm Component should update match the snapshot when there is more list items in the list state 1`] = `
<div
  className="Note"
>
  <div
    className="note-content"
    style={
      Object {
        "backgroundColor": "#FFE680",
      }
    }
  >
    <input
      autoComplete="off"
      className="note-title"
      name="title"
      onChange={[Function]}
      placeholder="Enter A Title..."
      type="text"
      value="Jake ToDo"
    />
    <div
      className="errorLabel"
    >
      <i
        className="fas fa-exclamation-triangle"
      />
      <p
        className="name-error"
      >
        Please Enter A Title
      </p>
    </div>
    <ul
      className="ListItems"
    >
      <ListItem
        handleItemChange={[Function]}
        handleItemDelete={[Function]}
        id="O6oYExG7Q"
        index={0}
        isComplete={false}
        key="O6oYExG7Q"
        text="Do basic styling"
        toggleComplete={[Function]}
      />
      <ListItem
        handleItemChange={[Function]}
        handleItemDelete={[Function]}
        id="1234552f4"
        index={1}
        isComplete={false}
        key="1234552f4"
        text="Javascript is cool"
        toggleComplete={[Function]}
      />
    </ul>
    <p
      className="Completed-Message"
    >
      1 Completed Item(s)
    </p>
    <ul
      className="ListItems Completed"
    >
      <ListItem
        handleItemChange={[Function]}
        handleItemDelete={[Function]}
        id="55ogax554"
        index={0}
        isComplete={true}
        key="55ogax554"
        text="Do More CsS"
        toggleComplete={[Function]}
      />
    </ul>
    <NoteOptions
      changeColor={[Function]}
      deleteNote={[Function]}
      displayError={false}
      handleClose={[Function]}
      handleType={[Function]}
      title="Jake ToDo"
      type="existing-note"
    />
  </div>
</div>
`;
